<a href="https://colab.research.google.com/github/JoshTorre/Prediction-of-Product-Sales/blob/main/Prediction_of_Product_Sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prediction of Product Sales
  - Author: Joshua Torre

## Project Overview

The goal of this project is to help the retailer understand the properties of products and outlets that play crucial roles in increasing sales.

## Load and Inspect Data

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Import panda library
import pandas as pd

In [3]:
# Load data
fname = '/content/drive/MyDrive/Github/sales_predictions_2023.csv'

# Create dataframe
df = pd.read_csv(fname)

In [4]:
# Inspect data with df.info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [5]:
# Check number of rows and columns with df.shape
df.shape

(8523, 12)

## Clean Data

### Cleaning object columns

In [24]:
# Inspect only object columns
pd.set_option('display.max_rows', None)
df.select_dtypes('object').head(10)

Item_Identifier Item_Fat_Content              Item_Type Outlet_Identifier  \
0           FDA15          Low Fat                  Dairy            OUT049   
1           DRC01          Regular            Soft Drinks            OUT018   
2           FDN15          Low Fat                   Meat            OUT049   
3           FDX07          Regular  Fruits and Vegetables            OUT010   
4           NCD19          Low Fat              Household            OUT013   
5           FDP36          Regular           Baking Goods            OUT018   
6           FDO10          Regular            Snack Foods            OUT013   
7           FDP10          Low Fat            Snack Foods            OUT027   
8           FDH17          Regular           Frozen Foods            OUT045   
9           FDU28          Regular           Frozen Foods            OUT017   

  Outlet_Size Outlet_Location_Type        Outlet_Type  
0      Medium               Tier 1  Supermarket Type1  
1      Medium               Tier 3  Supermarket Type2  
2      Medium               Tier 1  Supermarket Type1  
3         NaN               Tier 3      Grocery Store  
4        High               Tier 3  Supermarket Type1  
5      Medium               Tier 3  Supermarket Type2  
6        High               Tier 3  Supermarket Type1  
7      Medium               Tier 3  Supermarket Type3  
8         NaN               Tier 2  Supermarket Type1  
9         NaN               Tier 2  Supermarket Type1

#### Item Identifier

In [25]:
# Inspect if 'Item_Identifier' has any null values
df['Item_Identifier'].isna().sum()

0

In [27]:
# Inspect if 'Item_Identifier' has duplicated values
df['Item_Identifier'].duplicated().value_counts()

True     6964
False    1559
Name: Item_Identifier, dtype: int64

Duplication of Product ID or 'Item_Identifier' will be accepted, since same products are sold at various stores or supermarkets.

#### Item Fat Content

In [28]:
# Inspect if 'Item_Fat_Content' has any null values
df['Item_Fat_Content'].isna().sum()

0

In [29]:
# Inspect value counts of 'Item_Fat_Content'
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

It seems like there are inconsistencies in the categories of 'Item_Fat_Content'. Based on the Data Dictionary of the dataset, 'Item_Fat_Content' should be categorized either Low Fat or Regular.

In [33]:
# Change 'LF' to 'Low Fat'
df['Item_Fat_Content'] = df['Item_Fat_Content'].str.replace('LF', 'Low Fat')

In [32]:
# Change 'reg' to 'Regular'
df['Item_Fat_Content'] = df['Item_Fat_Content'].str.replace('reg', 'Regular')

In [34]:
# Change 'low fat' to 'Low Fat'
df['Item_Fat_Content'] = df['Item_Fat_Content'].str.replace('low fat', 'Low Fat')

In [35]:
# Inspect new value counts of 'Item_Fat_Content'
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

Data for 'Item_Fat_Content' is now consistent with having a categorical value of either Low Fat or Regular.

#### Item Type

In [38]:
# Inspect if 'Item_Type' has any null values
df['Item_Type'].isna().sum()

0

In [39]:
# Inspect value counts of 'Item_Type'
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

It seems like values  of 'Item_Type' are properly categorized.

#### Outlet Identifier

In [40]:
# Check value counts of 'Outlet_Identifier'
df['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [51]:
# Check if 'Outlet_Identifier' has any null values
df['Outlet_Identifier'].isna().sum()

0

It seems like there are not any changes to be made on the column of 'Outlet_Identifier'.

#### Outlet Size

In [53]:
# Check if 'Outlet_Size' has any null values
df['Outlet_Size'].isna().sum()

2410

In [56]:
# Inspect value counts of 'Outlet_Size'
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

The column of 'Outlet_Size' has 2,410 null values. These should be filled with either 'Small', 'Medium', or 'High' categorical value.

In [59]:
# Inspect null values of 'Outlet_Size'
df.loc[df['Outlet_Size'].isna()].head(10)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3            FDX07       19.200          Regular         0.000000   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
25           NCD06       13.000          Low Fat         0.099887   
28           FDE51        5.925          Regular         0.161467   
30           FDV38       19.250          Low Fat         0.170349   
33           FDO23       17.850          Low Fat         0.000000   
45           FDM39        6.420          Low Fat         0.089499   
46           NCP05       19.600          Low Fat         0.000000   
47           FDV49       10.000          Low Fat         0.025880   

                Item_Type  Item_MRP Outlet_Identifier  \
3   Fruits and Vegetables  182.0950            OUT010   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
25              Household   45.9060            OUT017   
28                  Dairy   45.5086            OUT010   
30                  Dairy   55.7956            OUT010   
33                 Breads   93.1436            OUT045   
45                  Dairy  178.1002            OUT010   
46     Health and Hygiene  153.3024            OUT045   
47                 Canned  265.2226            OUT045   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                        1998         NaN               Tier 3   
8                        2002         NaN               Tier 2   
9                        2007         NaN               Tier 2   
25                       2007         NaN               Tier 2   
28                       1998         NaN               Tier 3   
30                       1998         NaN               Tier 3   
33                       2002         NaN               Tier 2   
45                       1998         NaN               Tier 3   
46                       2002         NaN               Tier 2   
47                       2002         NaN               Tier 2   

          Outlet_Type  Item_Outlet_Sales  
3       Grocery Store           732.3800  
8   Supermarket Type1          1076.5986  
9   Supermarket Type1          4710.5350  
25  Supermarket Type1           838.9080  
28      Grocery Store           178.4344  
30      Grocery Store           163.7868  
33  Supermarket Type1          2174.5028  
45      Grocery Store           358.2004  
46  Supermarket Type1          2428.8384  
47  Supermarket Type1          5815.0972

### Cleaning number columns

In [ ]:
# Inspect only number columns
df.select_dtypes('number')

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0           9.300         0.016047  249.8092                       1999   
1           5.920         0.019278   48.2692                       2009   
2          17.500         0.016760  141.6180                       1999   
3          19.200         0.000000  182.0950                       1998   
4           8.930         0.000000   53.8614                       1987   
...           ...              ...       ...                        ...   
8518        6.865         0.056783  214.5218                       1987   
8519        8.380         0.046982  108.1570                       2002   
8520       10.600         0.035186   85.1224                       2004   
8521        7.210         0.145221  103.1332                       2009   
8522       14.800         0.044878   75.4670                       1997   

      Item_Outlet_Sales  
0             3735.1380  
1              443.4228  
2             2097.2700  
3              732.3800  
4              994.7052  
...                 ...  
8518          2778.3834  
8519           549.2850  
8520          1193.1136  
8521          1845.5976  
8522           765.6700  

[8523 rows x 5 columns]

In [ ]:
# Identify which columns needs to be manipulated using df.isna()
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
                             ... 
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
Length: 12, dtype: int64

Since columns 'Item_Weight' and 'Outlet_Size' have many null values, these columns will be our priority for data manipulation.

In [ ]:
# Inspect null values of 'Item_Weight'
df.loc[df['Item_Weight'].isna()]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In this dataframe, my strategy would be to imputate a value on 'Item_Weight' based on the mean of 'Item_Type'

In [ ]:
# Inspect columns 'Item_Weight' and 'Item_Type'
df[['Item_Weight', 'Item_Type']]

Item_Weight              Item_Type
0           9.300                  Dairy
1           5.920            Soft Drinks
2          17.500                   Meat
3          19.200  Fruits and Vegetables
4           8.930              Household
...           ...                    ...
8518        6.865            Snack Foods
8519        8.380           Baking Goods
8520       10.600     Health and Hygiene
8521        7.210            Snack Foods
8522       14.800            Soft Drinks

[8523 rows x 2 columns]

In [ ]:
# Inspect value counts for 'Item_Type'
pd.set_option('display.max_rows', None)
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

## Exploratory Data Analysis

## Feature Inspection